### PR TITLE
fix(packages): Make underline respect shrink/strech (rules package)

### LINKED ISSUE #1320: \underline and \boxaround do not honor stretcheability
SILE 0.12.1

The `\underline` command and the (undocumented) `\boxaround` command from the **rules** package do not respect shrinking/stretching (caused by an explicit glue or even line justification)

```
\begin{document}
\script[src=packages/rules]
\language[main=und]% Do disable hyphenation and have stretched line easily
\nofolios

Some \underline{underlined text}

Some \boxaround{boxed text}

Some \underline{underlined\hfill text}\hfill

Some \boxaround{boxed\hfill text}\hfill

Some \underline{test lorem ipsum} dor sit amet. Some test lorem ipsum dor sit amet. Some test lorem ipsum dor sit amet. 
Some ametloremipsum dor sit amet. Some test lorem ipsum dor sit amet. Some test lorem ipsum dor sit amet. 

Some \boxaround{test lorem ipsum} dor sit amet. Some test lorem ipsum dor sit amet. Some test lorem ipsum dor sit amet. 
Some ametloremipsum dor sit amet. Some test lorem ipsum dor sit amet. Some test lorem ipsum dor sit amet. 

\end{document}
```
image

I got the problem with my own framebox package and I am attempting a fix for it - if it works there, I could possibly try to check the rules package.

By the way, the `boxaround` command is not documented, should we even keep it (it lacks options - perhaps I could propose my framebox package for replacement, let me know...)

### PR DESCRIPTION
**Pull request by Omikhleia**
Closes #1320 

- Changed implementation to support shrinking/stretching in underlined text (`\underline` command).
- A fixed the undocumented `\boxaround` command for the same reason.
- Marked the latter as deprecated:
  - It was undocumented
  - It's pretty rough in what it does (no padding, no option, etc.) and unsatisfying
- Regarding underlining, the OpenType specs state that underlinePosition is the distance of _the top of the underline from the baseline_. I interpreted it strictly, i.e. the thickness is drawn below. The previous implementation lowered the line, but drew it upwards, which seems wrong to me.

Caveat: I did try RTL (arabic with Amiri font), and the commands do not do what I would have possibly expected. The older implementation was broken too, however, so this fix is not causing more harm...

### CODE REVIEW

File: packages/rules.lua [hunk 1]
```diff
@@ -43,25 +43,72 @@ SILE.registerCommand("underline", function (_, content)
...
+      -- should expand downwards
+      SILE.outputter:drawRule(oldX, Y + underlinePosition, newX - oldX, underlineThickness)
+    end
+  })
 end, "Underlines some content")
 
 SILE.registerCommand("boxaround", function (_, content)
+  -- This command was not documented and lacks feature.
+  -- Plan replacement with a better suited package.
+  SU.deprecated("\\boxaround (undocumented)", "\\framebox (package)", "0.12.0", "0.13.0")
```
Comment (alerque) [inline packages/rules.lua:77]: This bit needs splitting out into a different commit for the sake of the release notes. Deprecations stuffed inside `fix:` commits won't get the airtime they require.

File: packages/rules.lua [hunk 2]
```diff
@@ -43,25 +43,72 @@ SILE.registerCommand("underline", function (_, content)
   local underlineThickness = font.post.underlineThickness / upem * fontoptions.size
 
   local hbox = SILE.call("hbox", {}, content)
-  local gl = SILE.length() - hbox.width
-  SILE.call("lower", {height = SU.cast("measurement", underlinePosition)}, function()
-    SILE.call("hrule", {width = gl.length, height = underlineThickness})
-  end)
-  SILE.typesetter:pushGlue({width = hbox.width})
+  table.remove(SILE.typesetter.state.nodes) -- steal it back...
```
Comment (alerque) [inline packages/rules.lua:46]: Neither here no there for this PR, but I've thought several times this little "trick" needs to be upgraded to be a first class API of some kind, phantom mode or a reset command with mark points or a push/pop stack or something...

File: packages/rules.lua (hunk 1 above)
Comment (alerque) [reply packages/rules.lua:77]: P.S. I don't mind doing the split, I have `git revise -c` + interactive diff edits and am not afraid to use them. While we're in here I'd fix the underline and boxaround commands in two commits, again for the automatic release note benefits.

**Review by alerque — CHANGES_REQUESTED** [review]

**Review by alerque — APPROVED** [review]

File: packages/rules.lua (hunk 1 above)
Comment (Omikhleia) [reply packages/rules.lua:77]: You were quicker than I :)
Great, thanks !

Comment (Omikhleia) [reply packages/rules.lua:77]: Oops the commit messages say "borderbox" for actually "boxaround"

File: packages/rules.lua (hunk 2 above)
Comment (Omikhleia) [reply packages/rules.lua:46]: Thinking alike, as this is something we do in a lot of places... I'd think of something such as `SILE.helper.box.hbox(content)` that would do the same things as the "hbox" command except the final `table.insert` so one doesn't need the nearly immediate `table.remove` in these cases ;

Might look like a long name, but `SILE.findInTree` could be e.g. relocated to `SILE.helper.ast.findInTree`, where other goodies (e.g. my old #1196 from my initial try at SILE, lol) could also find a place, without polluting the global SILE namespace.

Comment (Omikhleia) [reply packages/rules.lua:46]: Altenately, we could have an `hbox:unpushed` command (build the box in the current context, but doesn't pushes it to the node queue). Less general but maybe clearer. (But that's beyond this PR indeed, just for the discussion)

Comment (alerque) [reply packages/rules.lua:46]: Opened an issue to track this someplace it won't get lost when I close this PR:‌https://github.com/sile-typesetter/sile/issues/1326

**Pull request merged**